mycpp/gc_dict.h

```diff
@@ -307,49 +275,119 @@ void Dict<K, V>::clear() {
 //   - SetAndIntern<V>(D, &string_key, value)
 //   This will enable duplicate copies of the string to be garbage collected
 template <typename K, typename V>
-int Dict<K, V>::position_of_key(K key) {
-  for (int i = 0; i < capacity_; ++i) {
-    int special = entry_->items_[i];  // NOT an index now
-    if (special == kDeletedEntry) {
-      continue;  // keep searching
+int Dict<K, V>::find_key_in_index(K key) const {
+  if (capacity_ == 0) {
+    return -1;
+  }
+
+  // Hash the key onto a slot in the index. If the first slot is occupied, probe
+  // until an empty one is found.
+  unsigned h = hash_key(key);
+  int init_bucket = h % capacity_;
+
+  // If there's a vacancy because of deletion on the probe path for this key, we
+  // should fill it before consuming a slot that has never been used. This
+  // should help keep the index somewhat compact.
+  int tombstone = -1;
+  for (int i = init_bucket; i < capacity_; ++i) {
+    int pos = entry_->items_[i];  // NOT an index now
+    DCHECK(pos < len_);
+    if (pos == kDeletedEntry) {
+      if (tombstone == -1) {
+        tombstone = pos;
+      }
+      continue;
     }
-    if (special == kEmptyEntry) {
-      return -1;  // not found
+    if (pos == kEmptyEntry) {
+      if (tombstone != -1) {
+        return tombstone;
+      }
+      return i;
     }
+    unsigned h2 = hash_key(keys_->items_[pos]);
+    if (h == h2 && keys_equal(keys_->items_[pos], key)) {
+      return i;
+    }
+  }
+
+  // Didn't find anything. Try wrapping around.
+  for (int i = 0; i < init_bucket; ++i) {
+    int pos = entry_->items_[i];  // NOT an index now
+    DCHECK(pos < len_);
+    if (pos == kDeletedEntry) {
+      if (tombstone == -1) {
+        tombstone = pos;
+      }
+      continue;
+    }
+    if (pos == kEmptyEntry) {
+      if (tombstone != -1) {
+        return tombstone;
+      }
+      return i;
+    }
+    unsigned h2 = hash_key(keys_->items_[pos]);
+    if (h == h2 && keys_equal(keys_->items_[pos], key)) {
+      return i;
+    }
+  }
+
+  return tombstone;
+}
+
+template <typename K, typename V>
+int Dict<K, V>::find_key(K key) const {
+  if (entry_ != nullptr) {
+    // Common case.
+    int pos = find_key_in_index(key);
+    if (pos == -1 || entry_->items_[pos] < 0) {
+      return -1;
+    }
+    return entry_->items_[pos];
+  }
+
+  // GlobalDict. Just scan.
+  for (int i = 0; i < len_; ++i) {
     if (keys_equal(keys_->items_[i], key)) {
       return i;
     }
   }
-  return -1;  // table is completely full?  Does this happen?
+
+  // Not found.
+  return -1;
 }
 
 template <typename K, typename V>
 void Dict<K, V>::set(K key, V val) {
-  int pos = position_of_key(key);
-  if (pos == -1) {  // new pair
-    reserve(len_ + 1);
+  DCHECK(obj_header().heap_tag != HeapTag::Global);
+  int pos = find_key_in_index(key);
+  if (pos == -1) {
+    // No room. Resize and see if we can find a slot.
+    for (int attempt = 0; attempt < 3; ++attempt) {
+      reserve((capacity_ ?: 1) * 2);
```

Comment:
comment on this formula?  It's doing something different the first time?

Reply:
Ah sure. These ternary expressions are so terse, it may just be better to explode this into an `if`. Basically, `Dict` is initialized with `capacity_ = 0`. In that case we just want to pretend like it was size 1 so we don't resize to zero...

Reply:
Another option could be to initialize dicts to some minimum size. Then this just becomes `reserve(capacity_ * 2);`. Is there a reason we init to zero right now?

Reply:
The existing `RoundCapacity()` method ended up being perfect for this